nanonameserver.py

```diff
@@ -0,0 +1,194 @@
+        self.use_thread = use_thread
+        self.left = None
+        self.right = None
+        self.udp = None
+        self.udp_address = None
+        self.tcp = None
+        self.tcp_address = None
+
+    def __enter__(self):
+        (self.left, self.right) = socket.socketpair()
+        # We're making the UDP socket now so it can be sent to by the
+        # caller immediately (i.e. no race with the listener starting
+        # in the thread).
+        if self.enable_udp:
+            self.udp = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, 0)
+            self.udp.bind((self.address, self.port))
+            self.udp_address = self.udp.getsockname()
+        if self.enable_tcp:
+            self.tcp = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
+            self.tcp.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            self.tcp.bind((self.address, self.port))
+            self.tcp.listen()
+            self.tcp_address = self.udp.getsockname()
+        if self.use_thread:
+            self.start()
+        return self
+
+    def __exit__(self, ex_ty, ex_va, ex_tr):
+        if self.left:
+            self.left.close()
+        if self.use_thread and self.is_alive():
+            self.join()
+        if self.right:
+            self.right.close()
+        if self.udp:
+            self.udp.close()
+        if self.tcp:
+            self.tcp.close()
+
+    async def wait_for_input_or_eof(self):
+        #
+        # This trio task just waits for input on the right half of the
+        # socketpair (the left half is owned by the context manager
+        # returned by launch).  As soon as something is read, or the
+        # socket returns EOF, EOFError is raised, causing a the
+        # nursery to cancel all other nursery tasks, in particular the
+        # listeners.
+        #
+        try:
+            with trio.socket.from_stdlib_socket(self.right) as sock:
+                self.right = None  # we own cleanup
+                await sock.recv(1)
+        finally:
+            raise EOFError
+
+    def handle(self, message):
+        #
+        # Handle message 'message'.  Override this method to change
+        # how the server behaves.
+        #
+        # The return value is either a dns.message.Message or a bytes.
+        # We allow a bytes to be returned for cases where handle wants
+        # to return an invalid DNS message for testing purposes.
+        #
+        r = dns.message.make_response(message)
+        r.set_rcode(dns.rcode.REFUSED)
+        return r
+
+    async def serve_udp(self):
+        with trio.socket.from_stdlib_socket(self.udp) as sock:
+            self.udp = None  # we own cleanup
+            while True:
+                try:
+                    (wire, from_address) = await sock.recvfrom(65535)
+                    q = dns.message.from_wire(wire)
+                    r = self.handle(q)
+                    if isinstance(r, dns.message.Message):
```

Comment:
It might be good to have a helper method that encapsulates the `handle()` call and the potential rendering shared by UDP and TCP processing, in case there are other extensions later.  The helper could also catch exceptions, which could be useful.